README
Crypt-Affine

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After  installing, you  can  find documentation for this module with  the perldoc
command.

    perldoc Crypt::Affine

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-Affine

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Crypt-Affine

    CPAN Ratings
        http://cpanratings.perl.org/d/Crypt-Affine

    Search CPAN
        http://search.cpan.org/dist/Crypt-Affine/

LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Crypt::Affine

0.11  2017-02-15T12:50:00+01:00
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.10  2017-02-01T12:25:00+01:00
      - Added key 'provides' to the Makefile.PL script.

0.09  2016-08-27T08:20:00+01:00
      - Updated Copyright year information.
      - Tidied up unit test script t/00-load.t.

0.08  2015-12-11T10:00:00+01:00
      - Added LICENSE file.

0.07  2015-01-16T15:25:00+01:00
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.06  2014-10-26T18:28:00+01:00
      - Rebuild the distribution using ExtUtils::MakeMaker v7.00 so that META files created with dependencies.

0.05  2014-10-25T23:55:25+01:00
      - Updated Changes file according to CPAN::Changes::Spec package.
      - Added unit test changes.t file.

0.04  2014-10-25T21:50:25+01:00
      - Restructured code using 'Moo'.
      - Added section 'Repository' to the pod document.

0.03  2014-10-25T20:30:00+01:00
      - Removed Build.PL file.
      - Tidied up MANIFEST file.

0.02  2011-08-22T14:15:16+01:00
      - Minor changes to pod.

0.01  2011-08-11T11:22:00+01:00
      - First version, released.

lib/Crypt/Affine.pm
package Crypt::Affine;

$Crypt::Affine::VERSION   = '0.11';
$Crypt::Affine::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Crypt::Affine - Interface to the Affine cipher.

=head1 VERSION

Version 0.11

=cut

use 5.006;
use autodie;
use Data::Dumper;
use Crypt::Affine::Params qw($Num $FilePath $ZeroOrOne);

use Moo;
use namespace::clean;

=head1 DESCRIPTION

The affine cipher is a type of mono alphabetic substitution cipher,  wherein each
letter in an alphabet is mapped to its numeric equivalent &  then encrypted using
a  simple  mathematical  function. It inherits the weaknesses of all substitution
ciphers. In  the  affine  cipher  the  letters of an alphabet of size m are first
mapped  to  the  integers in the range 0..m-1. It then uses modular arithmetic to
transform  the  integer  that  each  plaintext letter corresponds to into another
integer  that  correspond to a ciphertext letter.The function for encryption of a
single letter can be defined as below:

E(x) = (mx + r) % l

where  'l'  is the size of the alphabet and 'm' & 'r' are the key  of cipher. The
value 'm' must be choosen such that 'm' and 'l' are coprime.

Similarly the function for decryption of a single letter can be defined as below:

D(x) = (m ^ -1) (x - r) % l

where  (m ^ -1)  is the modular multiplicative inverse of 'm' modulo 'l'  and  it
satisfies the equation below:

m (m ^ -1) % l = 1

=cut

has  'm'       => (is => 'ro', isa => $Num,       required => 1 );
has  'r'       => (is => 'ro', isa => $Num,       required => 1 );
has  'reverse' => (is => 'ro', isa => $ZeroOrOne, default => sub { return 0; });
has  'source'  => (is => 'ro', isa => $FilePath);

=head1 CONSTRUCTOR

The constructor expects the following parameters as described below in the table:

    +----------+----------+-------------------------------------------------+
    | Key      | Required | Description                                     |
    +----------+----------+-------------------------------------------------+
    |  m       |    Yes   | Any positive number.                            |
    |  r       |    Yes   | Any positive number.                            |
    |  reverse |    No    | 0 or 1, depending whether to use reverse set of |
    |          |          | alphabets. Default is 0.                        |
    |  source  |    No    | Filename with complete path containing comma    |
    |          |          | separated list of alphabets. By default it uses |
    |          |          | A-Z, a-z.                                       |
    +----------+----------+-------------------------------------------------+

    use strict; use warnings;
    use Crypt::Affine;

    my $affine = Crypt::Affine->new({ m => 5, r => 8 });

=head1 METHODS

=head2 encrypt()

Encrypts the given string of alphabets ignoring any non-alphabets.

    use strict; use warnings;
    use Crypt::Affine;

    my $affine    = Crypt::Affine->new({ m => 5, r => 8 });
    my $original  = 'affine cipher';
    my $encrypted = $affine->encrypt($original);

    print "Original : [$original]\n";
    print "Encrypted: [$encrypted]\n";

=cut

sub encrypt {
    my ($self, $data) = @_;

    return unless defined $data;

    $self->_prepare() unless defined $self->{_a};
    my $encrypt = '';
    foreach (split //,$data) {
        (_unsupported($_))
        ?
        ($encrypt .= $_)
        :
        ($encrypt .= $self->_encrypt($_));
    }

    return $encrypt;
}

=head2 decrypt()

Decrypts the given string of alphabets ignoring any non-alphabets.

    use strict; use warnings;
    use Crypt::Affine;

    my $affine    = Crypt::Affine->new({ m => 5, r => 8 });
    my $original  = 'affine cipher';
    my $encrypted = $affine->encrypt('affine cipher');
    my $decrypted = $affine->decrypt($encrypted);

    print "Original : [$original]\n";
    print "Encrypted: [$encrypted]\n";
    print "Decrypted: [$decrypted]\n";

=cut

sub decrypt {
    my ($self, $data) = @_;

    return unless defined $data;

    $self->_prepare() unless defined $self->{_a};
    my $decrypt = '';
    foreach (split //,$data) {
        (_unsupported($_))
        ?
        ($decrypt .= $_)
        :
        ($decrypt .= $self->_decrypt($_));
    }

    return $decrypt;
}

#
#
# PRIVATE METHODS

sub _prepare {
    my ($self) = @_;

    my @data = ();
    my ($i, $j) = (1, 1);
    my ($a_, $z_, $l_, %_a, %_z, $_data);

    if (defined($self->{'source'}) && (-e $self->{'source'})) {
        local undef $/;
        open(my $IN, $self->{'source'});
        $_data = <$IN>;
        close($IN);

        chomp $_data;
        @data = split /\,/,$_data;
    }

    @data = ('a'..'z', 'A'..'Z') unless scalar(@data);
    $l_ = scalar(@data);
    foreach (@data) {
        $a_->{$_} = $i++;
        $z_->{$_} = ($l_ + 1) - $j++;
    }
    $self->{'r'} = $l_ if ($self->{'r'} > abs($l_));

    %_a = reverse %{$a_};
    %_z = reverse %{$z_};

    $self->{'a_'} = $a_;
    $self->{'z_'} = $z_;
    $self->{'l_'} = $l_;
    $self->{'_a'} = \%_a;
    $self->{'_z'} = \%_z;
}

sub _encrypt {
    my ($self, $char) = @_;

    my $i = (($self->{'m'} * $self->{'a_'}->{$char}) + $self->{'r'}) % $self->{'l_'};
    $i = $self->{'l_'} if ($i == 0);

    (defined($self->{'reverse'}) && ($self->{'reverse'}))
    ?
    return $self->{'_z'}->{$i}
    :
    return $self->{'_a'}->{$i};
}

sub _decrypt {
    my ($self, $char) = @_;

    my $i = 0;
    my $j = 0;

    (defined($self->{'reverse'}) && ($self->{'reverse'}))
    ?
    ($i = $self->{'z_'}->{$char})
    :
    ($i = $self->{'a_'}->{$char});

    $j = (_multiplier($self->{'m'}, $self->{'l_'}) * ($i - $self->{'r'})) % $self->{'l_'};
    $j = $self->{'l_'} if ($j == 0);

    return $self->{'_a'}->{$j};
}

sub _unsupported {
    my ($byte) = @_;

    return 1 if ($byte =~ /[\#\+\%\&\=\,\;\:\!\?\.\"\'\-\<\>\(\)\[\]\@\\\_\s]/);
    return 0;
}

sub _multiplier {
    my ($a, $m) = @_;

    $m = abs($m);
    $a = $a % $m;
    my ($b, $x, $y, $n) = ($m, 1, 0);

    while ($a != 0) {
        $n = int($b / $a);
        ($a, $b, $x, $y) = ($b - $n * $a, $a, $y - $n * $x, $x);
    }

    return $y % $m;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Crypt-Affine>

=head1 BUGS

Please report any bugs/feature requests to C<bug-crypt-affine at rt.cpan.org>  or
through the web interface at  L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Crypt-Affine>.
I will be notified & then you'll automatically be notified of progress on your bug
as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Crypt::Affine

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-Affine>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Crypt-Affine>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Crypt-Affine>

=item * Search CPAN

L<http://search.cpan.org/dist/Crypt-Affine/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Crypt::Affine


lib/Crypt/Affine/Params.pm
package Crypt::Affine::Params;

$Crypt::Affine::Params::VERSION   = '0.11';
$Crypt::Affine::Params::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Crypt::Affine::Params - Placeholder for parameters for Crypt::Affine.

=head1 VERSION

Version 0.11

=cut

use 5.006;
use strict; use warnings;
use Data::Dumper;

use vars qw(@ISA @EXPORT @EXPORT_OK);

require Exporter;
@ISA = qw(Exporter);
@EXPORT_OK = qw($FilePath $ZeroOrOne $Num);

our $ZeroOrOne = sub  { die "ERROR: Invalid data found [$_[0]]" unless check_zero_or_one($_[0]); };
sub check_zero_or_one { return (defined($_[0]) && ($_[0] =~ /^\d$/) && ($_[0] == 0 || $_[0] == 1)); }

our $Num = sub { return check_num($_[0]); };
sub check_num  { die "ERROR: Invalid NUM data type [$_[0]]" unless (defined $_[0] && $_[0] =~ /^\d+$/); }

our $FilePath = sub { die "ERROR: Invalid file path [$_[0]]" unless check_file_path($_[0]); };
sub check_file_path { return (-f $_[0]) };

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Crypt-Affine>

=head1 BUGS

Please report any bugs/feature requests to C<bug-crypt-affine at rt.cpan.org>  or
through the web interface at  L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Crypt-Affine>.
I will be notified & then you'll automatically be notified of progress on your bug
as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Crypt::Affine::Params

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-Affine>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Crypt-Affine>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Crypt-Affine>

=item * Search CPAN

L<http://search.cpan.org/dist/Crypt-Affine/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Crypt::Affine::Params


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Crypt::Affine',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Crypt/Affine.pm',
    ABSTRACT_FROM      => 'lib/Crypt/Affine.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES   => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'autodie'          => 0,
        'Moo'              => '2.000000',
        'namespace::clean' => '0.24',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Crypt-Affine-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Crypt::Affine'         => { file => 'lib/Crypt/Affine.pm'       , version => '0.11' },
            'Crypt::Affine::Params' => { file => 'lib/Crypt/Affine/Params.pm', version => '0.11' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Crypt-Affine.git',
                web  => 'https://github.com/manwar/Crypt-Affine',
            },
        }})
     : ()
    ),
);
